src/script_opt/ZAM/README.md

```diff
@@ -88,16 +90,119 @@ issues:
-|`profile-ZAM`	|	Generate to _stdout_ a ZAM execution profile. (Requires configuring with `--enable-debug`.)|
+|`profile-ZAM`	|	Generate to "zprof.log" a ZAM execution profile. (Requires configuring with `--enable-ZAM-profiling` or `--enable-debug`.)|
 |`report-recursive`	|	Report on recursive functions and exit.|
 |`report-uncompilable`	|	Report on uncompilable functions and exit. For ZAM, all functions should be compilable.|
 |`xform`		|	Transform scripts to "reduced" form.|
 
+<a name="ZAM-profiling"></a>
+## ZAM Profiling
+
+ZAM supports detailed script execution profiling, activated using `-O
+profile-ZAM`. (This option implies `-O ZAM` unless you've already specified
+some ZAM optimization options.) Profiles are written to `zprof.log`. These
+profiles have a number of components, and are intended to be subsetted
+(such as by using `grep`) and then further processed, such as by using
+`sort` to pick out instances with large values.
+
+When profiling, ZAM gathers for each function body its total number of
+calls, total CPU time (including time spent in any script functions or
+BiFs it calls), an estimate of total memory allocations, and the number
+of sampled ZAM instructions (see below). Memory impact is done by comparing
+Zeek's total memory usage when a function body starts with its total usage
+when the body finishes. Any increase is charged against the function body;
+decreases however are not (because it's not as clear whether they are
+meaningful). This approach only approximates actual memory usage because
+often allocations can be met by memory currently available at user level,
+not requiring any kernel allocations; and if they _do_ require kernel
+allocations, those can be significantly larger than the immediate need.
+That said, experience finds that the resulting values (reported in bytes)
+do generally reflect the execution's state-holding impact.
+
+Often the report will state that a function body "did not execute" when
+in fact you're sure it did. This arises due to ZAM's heavy use of _inlining_:
+while the code of the function body did indeed execute, it did so only in
+the context of _other_ function bodies, and that's where the CPU & memory
+impact are charged. (You can suppress this by using `-O no-inline` to turn
+off ZAM's inlining, although with a significant performance impact.)
+
+In addition to per-connection profiling, ZAM also profiles individual ZAM
+instructions. Because fine-grained profiling of every instruction execution
+imposes a significant performance penalty,
+ZAM does instruction-level profiling using _sampling_.  The default sampling
+rate is 1-in-100. (You can control it via the `ZEEK_ZAM_PROF_SAMPLING_RATE`
+environment variable, and setting that variable to `1` effectively turns
+off sampling). More frequent sampling rates slow down execution further but
+provide more accurate information. With the default rate, the slowdown is
+about 2x, so not something to use in production in its present form.
```

Comment:
All nice thoughts. The other thing I've been thinking about is having a separate flag for only doing the flame-graph-relevant instrumentation (which can be sampled) and skipping the summary-per-function stats, which aren't germane for flame graphs and can't be safely sampled (because some rare-but-expensive functions can get entirely skipped, masking their cost, or magnifying it unduly if they happen to get sampled)